intersect-resources: Support prerender mode, patch for stale rects

choumx opened the pull request:
Partial for #25428.

- Keep using a single observer at 3vp and compute intersection with 1vp. In prerender mode, queue intersections outside of 1vp.
- Addresses the stale layout box issue by reverting to continue to use the current system i.e. requesting measure on mutations, scheduling passes after element size changes, etc.
- Removes some of the edge case handling for simplicity. I'll restore them if needed later.

Will add tests next.

/to @dvoytenko @jridgewell /cc @ampproject/wg-runtime

samouri commented on src/service/resources-impl.js:
```diff
@@ -281,19 +284,61 @@ export class ResourcesImpl {
+        e => this.intersectsThreeViewports_(e),
+        {
+          root: this.oneViewportObserver_.root,
+          rootMargin: '200% 25%',
```
why is the margin width 25%?

choumx replied:
That's the current size of our loading rectangle (`loadRect` I think). One example is animating elements that start just off-screen, horizontally.

samouri replied:
Shouldn't we also have a `rootMargin: '0% 25%'` for the `oneViewportObserver_ ` then?

choumx replied:
`loadRect` in prerender mode doesn't have it. Also I think we'd want to deprioritize such elements relative to 1vp.

samouri replied:
ah, i see now. makes sense!

samouri commented on src/service/resources-impl.js:
```diff
@@ -331,16 +382,14 @@ export class ResourcesImpl {
-        const wasIntersecting = r.isInViewport();
         let isDisplayed = this.measureResource_(r, clientRect);
-
         if (wasIntersecting && !isIntersecting) {
```
in which situation would `isDisplayed` and `isIntersecting` have different values.

choumx replied:
`isDisplayed()` reflects size of the bounding rectangle. It should probably be renamed. Happy to do that separately.

samouri replied:
ah, like hidden vs. visible. the name isn't bad, but when it comes to viewport detection "displayed" is definitely overloaded

samouri commented on src/service/resources-impl.js:
```diff
@@ -657,8 +712,11 @@ export class ResourcesImpl {
-    if (this.intersectionObserver_) {
-      this.intersectionObserver_.unobserve(resource.element);
+    if (this.oneViewportObserver_) {
+      this.oneViewportObserver_.unobserve(resource.element);
```
Are there times where we call `removeResource` without having the element also removed from the DOM? If so, why do we need to manually call `unobserve`, I'd expect IntersectionObserver to handle a removed element gracefully.

choumx replied:
Check out the callers of `Resources.remove()`: https://github.com/ampproject/amphtml/blob/6b439aec75c1d9b52ba19435f6730b27a457bf42/src/custom-element.js#L949-L950

samouri commented on src/service/resources-impl.js:
```diff
@@ -281,19 +284,61 @@ export class ResourcesImpl {
+          rootMargin: '200% 25%',
+        }
+      );
+      const elements = this.resources_.map(r => r.element);
```
Unsure if this is desirable, but would it be technically possible to have only a single 3p IO, and use the data in entries to calculate whether or not something intersects at viewport=1?

e.g.
```js
// takes in an entry from an IntersectionObserver with rootMargin 200% and determines if the item also intersects at 0%.
intersectsOneViewport(entryFrom3vpIO) {
  const {rootBounds, boundingClientRect, isIntersecting} = entryFrom3vpIO;
  if (!isIntersecting) {
    return false;
  }
  const vp1YTop= rootBounds.y + rootBounds.height/3;
  const vp1YBottom= rootBounds.y + rootBounds.height/3;
  const isWithinViewport = boundingClientRect.y >= vp1YTop && boundingClientRect.y <= vp1YBottom;

  // we don't care about X, since if it wasn't within our desired x-margin it would have early returned with isIntersecting=false. 
  return isWithintViewportY;
}
```

choumx replied:
Yea I think it's possible but less desirable. :)

choumx commented:
@jridgewell Friendly ping.

dvoytenko commented on src/service/resources-impl.js:
```diff
@@ -193,8 +193,17 @@ export class ResourcesImpl {
 
-    /** @private {?IntersectionObserver} */
-    this.intersectionObserver_ = null;
+    /**
```
The bigger question I have: if this continues to diverge in major ways, would it be easier to just fork this whole manager for the InOb experiment?

choumx replied:
Good question. I only have a few small changes left to do, so this is nearly the final state.

Also I'm concerned about bundle size since we can't leverage module/nomodule or other tricks easily.

dvoytenko commented on src/service/resources-impl.js:
```diff
@@ -281,94 +287,144 @@ export class ResourcesImpl {
+      for (let i = 0; i < this.resources_.length; i++) {
+        const r = this.resources_[i];
+        if (r.getState() == ResourceState.NOT_BUILT && !r.isBuilding()) {
+          this.buildOrScheduleBuildForResource_(r, /* checkForDupes */ true);
```
We are blocking build? We've not blocked build before - only loading. This is certain to be positive, but might have lots of regressions with the elements that do some early work on build?

Also, does the `whenBuilt` never fail for any individual elements, or all together? Overall this seems like a pretty strict rule. What if a 1vp element is very slow in download?

choumx replied:
Yea, I was originally planning to split this out but I think this should be safe to do since we already do something similar in prerender mode with our quota system. 

Good points on being bottlenecked by slow outliers. I'll think about how to mitigate.

kristoferbaxter commented on src/service/resources-impl.js:
```diff
@@ -281,94 +287,144 @@ export class ResourcesImpl {
+          rootMargin: '200% 25%',
+        }
+      );
+      const elements = this.resources_.map(r => r.element);
```
Is there value in doing a `map` then a `forEach`? Or does this indirection get removed by Closure Compiler?

```javascript
this.resources.forEach(r => (
  e.element.whenUpgraded().then(() => 
    this.threeViewportsObserver_.observe(r.element)
  )
);
```

choumx replied:
This used to be more complex -- I'll simplify. 👍

choumx commented:
> I think the handling of the two observers is weird. Why not just register both, and record what the actual distance to viewport is? Make you decisions from there. 

Heh, I assume you're talking about the lazy instantiation of 3vp observer? After observing an element, there's an initial intersection entry/callback (even if not intersecting). So lazy instantiation (1) allows the browser to defer the work and (2) avoids the need for us to store/queue the entry.

Alternatively, we could try just using one observer and always compute viewport distance. But this trades the complexity of two observers for a more complex handling with one observer. It's not clear to me that it's better, though I can try it out.

choumx commented:
Changed back to use one observer. Much simpler code, though at the cost of a few extra calculations and potentially staler rects due to queueing during prerender. 

Also removed the deferral of build on upgrade (will do that separately).

choumx commented:
@dvoytenko @jridgewell PTAL, the diff is much less scary now. :)

In other news, the spec change was supposed to land in Chrome 81 which has now been postponed unfortunately. The experiment can still activate on non-iframed pages.

dvoytenko commented on src/service/resources-impl.js:
```diff
@@ -286,89 +292,80 @@ export class ResourcesImpl {
-  intersects_(entries, unusedObserver) {
+  intersects_(entries) {
+    // TODO(willchou): Remove assert once #27167 is fixed.
+    devAssert(this.prerenderSize_ == 1);
```
What's the value of the `devAssert()` for this?

choumx replied:
Mostly as an FYI for readers. Assert would trigger for the few people that are playing with `prerenderSize` in development (maybe just me).

choumx commented:
Update: discovered a preexisting race condition (#27483) that affects the current IntersectionObserver implementation. Fixing this race is easy but I'm leaning towards a lower impact/risk integration that keeps most of the `discoverWork_` codepath intact for MVP. 

Closing this PR for now.
